### PR TITLE
DRAFT: Adding files feature

### PR DESCRIPTION
**Pull request by Euphoride**
Solves #21, still draft, welcoming reviews keeping in mind TODOs need to be addressed.

### CODE REVIEW

File: frontend/tsconfig.json [hunk 1]
```diff
@@ -14,7 +14,7 @@
     "resolveJsonModule": true,
     "isolatedModules": true,
     "noEmit": true,
-    "jsx": "react-jsx"
```
Comment (NathanielB123) [inline frontend/tsconfig.json]: Got thrown by this as well in https://github.com/gpuctl/gpuctl/pull/239. I *think* if you `npm update` then prettier will change it's mind and want the trailing commas gone.

**Review by YRMYJASKA — COMMENTED** [review]
LGTM on the serverside stuff, cannot comment on the react code at all

File: frontend/src/Components/GroupInfoManagement.tsx [hunk 2]
```diff
@@ -59,6 +76,110 @@ export const GroupInfoManagement = ({
...
+              const downloadLink = document.createElement("a");
+              downloadLink.href = URL.createObjectURL(
+                new Blob([await r.arrayBuffer()]),
+              );
+              downloadLink.download = currentFile;
+              downloadLink.click();
+            }
+          },
+          failure: () => {
+            // alert("Failure");
```
Comment (NathanielB123) [inline frontend/src/Components/GroupInfoManagement.tsx:99]: Is there a chance we could create a `toast` here to inform the user that something went wrong? (see `useAddMachine` in `Hooks/Hooks.tsx` for an example)

EDIT: Oops, forgot the semantics of `useAuthFetch`. An invalid response is still considered a `success` so the toast should be above (in the case the `status != 200` but in any case - `toast`s on failed requests are probably a good idea.

EDIT 2: Actually, might be better actually to put the toast logic higher up in `Hooks.tsx` (consistent with `useAddMachine` etc... and means that if we use the hook in multiple places the toast logic doesn't need to be duplicated).

File: frontend/src/Components/GroupInfoManagement.tsx [hunk 3]
```diff
@@ -59,6 +76,110 @@ export const GroupInfoManagement = ({
...
+    const downloader = useGetSpecificFile(
+      currentMachine,
+      currentFile,
+      (response) => {
+        validatedElim(response, {
+          success: async (r) => {
+            // here the response should have given an attached file through the attachment header stuff
+            // let's download
+
+            if (r.status == 200) {
```
Comment (NathanielB123) [inline frontend/src/Components/GroupInfoManagement.tsx:89]: NIT: Elsewhere in the codebase we are using `===`

File: frontend/src/Components/GroupInfoManagement.tsx [hunk 4]
```diff
@@ -59,6 +76,110 @@ export const GroupInfoManagement = ({
...
+    };
+  };
+
+  useEffect(() => {
+    if (currentMachine) {
+      view();
+    }
+  }, [currentMachine]);
+
+  // ...
```
Comment (NathanielB123) [inline frontend/src/Components/GroupInfoManagement.tsx:149]: NIT: What does this comment mean?

File: frontend/src/Hooks/Hooks.tsx [hunk 5]
```diff
@@ -2,14 +2,95 @@ import { useToast } from "@chakra-ui/react";
...
+  const len = file.byteLength;
+  for (let i = 0; i < len; i++) {
+    binaryString += String.fromCharCode(file[i]);
+  }
+  return btoa(binaryString);
+};
+
+export const useUploadFile = (callback: (r: Validated<Response>) => void) => {
+  const { useAuthFetch } = useAuth();
+  const [, uploadFileAuth] = useAuthFetch(UPLOAD_FILE_URL, callback);
```
Comment (NathanielB123) [inline frontend/src/Hooks/Hooks.tsx:48]: Toast?

File: frontend/src/Hooks/Hooks.tsx [hunk 6]
```diff
@@ -2,14 +2,95 @@ import { useToast } from "@chakra-ui/react";
...
+        mime,
+        filename,
+        file_enc: encodeFile(file),
+      }),
+    });
+};
+
+export const useRemoveFile = () => {
+  const { useAuthFetch } = useAuth();
+  const [, removeFileAuth] = useAuthFetch(REMOVE_FILE_URL);
```
Comment (NathanielB123) [inline frontend/src/Hooks/Hooks.tsx:67]: Toast?

File: frontend/src/Hooks/Hooks.tsx [hunk 7]
```diff
@@ -2,14 +2,95 @@ import { useToast } from "@chakra-ui/react";
...
+
+export const useGetSpecificFile = (
+  hostname: string,
+  filename: string,
+  callback: (r: Validated<Response>) => void,
+) => {
+  const { useAuthFetch } = useAuth();
+  const [, getSpecificFileAuth] = useAuthFetch(
+    GET_SPECIFIC_FILE_URL + "?hostname=" + hostname + "&file=" + filename,
+    callback,
```
Comment (NathanielB123) [inline frontend/src/Hooks/Hooks.tsx:86]: Toast?

**Review by NathanielB123 — APPROVED** [review]
Happy to merge now if you want. Just a couple nitpicks and ideally there would be some feedback to the user if the requests fail (`toast`s), but that can probably be addressed in a follow-up PR (i.e. if we want to try and merge this quickly to test in deployment).

**Pull request merged**